src/pubnub.rs

```diff
@@ -0,0 +1,491 @@
+    ///
+    /// while let Some(message) = stream.next().await {
+    ///     println!("Received message: {:?}", message);
+    /// }
+    /// # };
+    /// ```
+    pub async fn subscribe(&mut self, channel: &str) -> Subscription {
+        let (channel_tx, mut channel_rx) = mpsc::channel(10);
+
+        // Hold the lock for the entire duration of this function
+        let mut guard = self.pipe.lock().await;
+
+        let id = if let Some(pipe) = guard.as_mut() {
+            // Send an "add channel" message to the subscribe loop
+            let channel = ListenerType::Channel(channel.to_string());
+            debug!("Adding channel: {:?}", channel);
+
+            let result = pipe
+                .tx
+                .send(PipeMessage::Add(channel, channel_tx.clone()))
+                .await;
+
+            if result.is_ok() {
+                // Fetch id from `SubscribeLoop`
+                // Uses `channel_rx` which is unique to each `Subscription`.
+                let msg = channel_rx.next().await;
+
+                if let Some(Message {
+                    message_type: Type::Ready(id),
+                    ..
+                }) = msg
+                {
+                    id
+                } else {
+                    panic!("Unexpected message: {:?}", msg);
+                }
+            } else {
+                // When sending to the pipe fails, recreate the SubscribeLoop
+                *guard = None;
+
+                0
+            }
+        } else {
+            0
+        };
+
+        if guard.is_none() {
+            // Create communication pipe
+            let (my_pipe, their_pipe) = {
+                let (my_tx, their_rx) = mpsc::channel(10);
+                let (their_tx, my_rx) = mpsc::channel(10);
+
+                let my_pipe = Pipe {
+                    tx: my_tx,
+                    rx: my_rx,
+                };
+                let their_pipe = Pipe {
+                    tx: their_tx,
+                    rx: their_rx,
+                };
+
+                (my_pipe, their_pipe)
+            };
+            *guard = Some(my_pipe);
+
+            let mut channels: ChannelMap = ChannelMap::new();
+            let listeners = channels
+                .entry(channel.to_string())
+                .or_insert_with(Default::default);
+            listeners.push(channel_tx);
+
+            // Create subscribe loop
+            debug!("Creating SubscribeLoop");
+            let subscribe_loop = SubscribeLoop::new(
+                their_pipe,
+                self.client.clone(),
+                self.origin.clone(),
+                self.agent.clone(),
+                self.subscribe_key.clone(),
+                channels,
+                ChannelMap::new(),
+            );
+
+            // Spawn the subscribe loop onto the Tokio runtime
+            tokio::spawn(subscribe_loop.run());
+
+            debug!("Waiting for long-poll...");
+            guard
```

Comment:
Hmmm why is this necessary -- do we even need the Ready message?

Reply:
You're right, the `Ready` message is no longer needed. It was in place before the mutex was added. With that, the `mpsc` channel bound can also be set to 1. Will make the adjustments.

Reply:
oooh, I am completely mistaken. We can't remove the ready event for two reasons:

1. In the case that a `SubscribeLoop` already exists, this message will carry the `id` necessary for dropping the stream.
2. In the other case where a new `SubscribeLoop` needs to be created, this test (and others like it) will fail without waiting: https://github.com/stephenlb/pubnub-rust/blob/50081622f6e2caa496836ea8516bfb536508bd88/src/lib.rs#L127-L143

In the latter case, we can move the wait-for-ready signal to `#[cfg(test)]` at the cost of missing an opportunity to signal to the listener that the stream is actually capable of receiving messages immediately.

Reply:
Well, not _completely_ mistaken! We can still drop the `mpsc` bound to 1.